src/devices/efergy_optical.c

```diff
@@ -0,0 +1,109 @@
+/* Efergy Optical (electricity meter)
+ *
+ * This electricity meter periodically reports current power consumption
+ * on frequency ~433.55 MHz. The data that is transmitted consists of 8
+ * bytes:
+ *
+ * Byte 1-4: Start bits (0000), then static data (probably device id)
+ * Byte 5-7: all zeros 
+ * Byte 8: Pulse Count
+ * Byte 9: Unknown (value is 15 but 16 for low pulses)
+ * Byte 10: seconds
+ * Byte 11: crc16 xmodem
+ * Byte 12: crc16 xmodem
+ * if pulse count <3 then energy =(( pulsecount/impulse-perkwh) * (3600/seconds))
+ * else energy = ((pulsecount/n_imp) * (3600/seconds))
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+#include "rtl_433.h"
+#include "util.h"
+#include "data.h"
+
+static int efergy_optical_callback(bitbuffer_t *bitbuffer) {
+	unsigned num_bits = bitbuffer->bits_per_row[0];
+	uint8_t *bytes = bitbuffer->bb[0];
+	float energy;
+	int pulsecount, n_imp;
+	int seconds;
+	data_t *data;
+        char time_str[LOCAL_TIME_BUFLEN];
+
+	if (num_bits < 64 || num_bits > 100) {
+		return 0;
+	}
+
+	// The bit buffer isn't always aligned to the transmitted data, so
+	// search for data start and shift out the bits which aren't part
+	// of the data. The data always starts with 0000 (or 1111 if
+	// gaps/pulses are mixed up).
+	while ((bytes[0] & 0xf0) != 0xf0 && (bytes[0] & 0xf0) != 0x00) {
+		num_bits -= 1;
+		if (num_bits < 64) {
+			return 0;
+		}
+
+		for (unsigned i = 0; i < (num_bits + 7) / 8; ++i) {
+			bytes[i] <<= 1;
+			bytes[i] |= (bytes[i + 1] & 0x80) >> 7;
+		}
+	}
+
+	// Sometimes pulses and gaps are mixed up. If this happens, invert
+	// all bytes to get correct interpretation.
+	if (bytes[0] & 0xf0) {
+		for (unsigned i = 0; i < 11; ++i) {
+			bytes[i] = ~bytes[i];
+		}
+	}
+
```

Comment:
This code is copied from the efergy e2 decoder.

For the efergy e2 the fm demodulation/decoding is much more robust these days (probably because of the fm demodulator improvements recently) so it seems it's no longer true that the pulse trains sometimes get inverted.

But, today it looks like the if-statement always evaluates to true (for the efergy e2 test data in rtl_433_tests) and the data always needs to be inverted for the following decoding to work. The reason for this seems to be that the signal produced by the efergy e2 doesn't quite match what the pwm decoder expects (see http://electrohome.pbworks.com/w/page/34379858/Efergy%20Elite%20Wireless%20Meter%20Hack for pulse details).

Do you see any way to better handle the efergy signal?